pkg/controller/connectivitycheck/connectivity_check_controller.go

```diff
@@ -0,0 +1,277 @@
+	c := ovnKubernetesConnectivityCheckController{
+		ConnectivityCheckController: connectivitycheckcontroller.NewConnectivityCheckController(
+			"openshift-ovn-kubernetes",
+			operatorClient,
+			operatorcontrolplaneClient,
+			apiextensionsClient,
+			apiextensionsInformers,
+			[]factory.Informer{
+				operatorClient.Informer(),
+				kubeInformersForNamespaces.InformersFor("openshift-ovn-kubernetes").Core().V1().Pods().Informer(),
+				kubeInformersForNamespaces.InformersFor("openshift-kube-apiserver").Core().V1().Endpoints().Informer(),
+				kubeInformersForNamespaces.InformersFor("openshift-kube-apiserver").Core().V1().Services().Informer(),
+				configInformers.Config().V1().Infrastructures().Informer(),
+			},
+			recorder,
+			true,
+		),
+	}
+	generator := &connectivityCheckTemplateProvider{
+		operatorClient:             operatorClient,
+		operatorcontrolplaneClient: operatorcontrolplaneClient,
+		endpointsLister:            kubeInformersForNamespaces.InformersFor("openshift-kube-apiserver").Core().V1().Endpoints().Lister(),
+		serviceLister:              kubeInformersForNamespaces.InformersFor("openshift-kube-apiserver").Core().V1().Services().Lister(),
+		podLister:                  kubeInformersForNamespaces.InformersFor("openshift-ovn-kubernetes").Core().V1().Pods().Lister(),
+		infrastructureLister:       configInformers.Config().V1().Infrastructures().Lister(),
+	}
+	return c.WithPodNetworkConnectivityCheckFn(generator.generate)
+}
+
+type ovnKubernetesConnectivityCheckController struct {
+	connectivitycheckcontroller.ConnectivityCheckController
+}
+
+type connectivityCheckTemplateProvider struct {
+	operatorClient             v1helpers.OperatorClient
+	operatorcontrolplaneClient *operatorcontrolplaneclient.Clientset
+	endpointsLister            corev1listers.EndpointsLister
+	serviceLister              corev1listers.ServiceLister
+	podLister                  corev1listers.PodLister
+	infrastructureLister       configv1listers.InfrastructureLister
+}
+
+func (c *connectivityCheckTemplateProvider) generate(ctx context.Context, syncContext factory.SyncContext) ([]*v1alpha1.PodNetworkConnectivityCheck, error) {
+	var templates []*v1alpha1.PodNetworkConnectivityCheck
+	// kas service IP
+	templates = append(templates, c.getTemplatesForKubernetesServiceMonitorService(syncContext.Recorder())...)
+	// kas default service IP
+	templates = append(templates, c.getTemplatesForKubernetesDefaultServiceCheck(syncContext.Recorder())...)
+	// each kas endpoint IP
+	templates = append(templates, c.getTemplatesForKubernetesServiceEndpointsChecks(syncContext.Recorder())...)
+	// each api load balancer hostname
+	templates = append(templates, c.getTemplatesForApiLoadBalancerChecks(syncContext.Recorder())...)
+
+	pods, err := c.podLister.List(labels.Set{"app": "ovnkube-node"}.AsSelector())
+	if err != nil {
+		syncContext.Recorder().Warningf("EndpointDetectionFailure", "failed to list openshift-apiserver pods: %v", err)
```

Comment:
error message doesn't match what the code is doing

Reply:
done